teamcity: failed tests on master: testrace/TestRaftSSTableSideloadingUpdatedReplicaID, test/TestRaftSSTableSideloadingUpdatedReplicaID

The following tests appear to have failed:

[#762920](https://teamcity.cockroachdb.com/viewLog.html?buildId=762920):


```
--- FAIL: testrace/TestRaftSSTableSideloadingUpdatedReplicaID (0.000s)
Race detected!

------- Stdout: -------
I180706 21:01:03.982150 11752 storage/engine/rocksdb.go:566  opening rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID265162223"
I180706 21:01:04.059144 11752 storage/replica_sideload_test.go:1040  olddir is /go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID265162223/auxiliary/sideloading/1/1.1, newdir is /go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID265162223/auxiliary/sideloading/1/1.2
I180706 21:01:04.059356 11752 storage/engine/rocksdb.go:699  closing rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID265162223"
I180706 21:01:04.060610 11752 util/stop/stopper.go:471  quiescing; tasks left:
1      [async] gossip.infoStore: callback
==================
Read at 0x00c420572560 by goroutine 43:
  github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).Iterate()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:798 +0x57
  github.com/cockroachdb/cockroach/pkg/storage/compactor.(*Compactor).examineQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/compactor/compactor.go:421 +0x234
  github.com/cockroachdb/cockroach/pkg/storage/compactor.(*Compactor).Start.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/compactor/compactor.go:131 +0x3f9
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:193 +0x14a

Previous write at 0x00c420572560 by goroutine 32:
  github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).Close()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:712 +0x297
  github.com/cockroachdb/cockroach/pkg/storage.TestRaftSSTableSideloadingUpdatedReplicaID()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_sideload_test.go:1048 +0x7ef
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 43 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:186 +0xba
  github.com/cockroachdb/cockroach/pkg/storage/compactor.(*Compactor).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/compactor/compactor.go:110 +0x13c
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1460 +0xa05
  github.com/cockroachdb/cockroach/pkg/storage.(*testContext).StartWithStoreConfig()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_test.go:208 +0x1333
  github.com/cockroachdb/cockroach/pkg/storage.(*testContext).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_test.go:158 +0x29d
  github.com/cockroachdb/cockroach/pkg/storage.TestRaftSSTableSideloadingUpdatedReplicaID()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_sideload_test.go:1004 +0x1f7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 32 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:978 +0x2cd
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/main_test.go:57 +0x2a8
  main.main()
      _testmain.go:1062 +0x22a
==================
--- FAIL: test/TestRaftSSTableSideloadingUpdatedReplicaID (0.060s)

------- Stdout: -------
I180706 20:49:26.935853 12032 storage/engine/rocksdb.go:566  opening rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID081132010"
I180706 20:49:26.972385 12032 storage/store.go:1449  [s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180706 20:49:26.985020 12032 storage/replica_sideload_test.go:1040  olddir is /go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID081132010/auxiliary/sideloading/1/1.1, newdir is /go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID081132010/auxiliary/sideloading/1/1.2
I180706 20:49:26.985063 12032 storage/engine/rocksdb.go:699  closing rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID081132010"Race detected!

------- Stdout: -------
I180706 21:01:03.982150 11752 storage/engine/rocksdb.go:566  opening rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID265162223"
I180706 21:01:04.059144 11752 storage/replica_sideload_test.go:1040  olddir is /go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID265162223/auxiliary/sideloading/1/1.1, newdir is /go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID265162223/auxiliary/sideloading/1/1.2
I180706 21:01:04.059356 11752 storage/engine/rocksdb.go:699  closing rocksdb instance at "/go/src/github.com/cockroachdb/cockroach/artifacts/TestRaftSSTableSideloadingUpdatedReplicaID265162223"
I180706 21:01:04.060610 11752 util/stop/stopper.go:471  quiescing; tasks left:
1      [async] gossip.infoStore: callback
==================
Read at 0x00c420572560 by goroutine 43:
  github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).Iterate()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:798 +0x57
  github.com/cockroachdb/cockroach/pkg/storage/compactor.(*Compactor).examineQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/compactor/compactor.go:421 +0x234
  github.com/cockroachdb/cockroach/pkg/storage/compactor.(*Compactor).Start.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/compactor/compactor.go:131 +0x3f9
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:193 +0x14a

Previous write at 0x00c420572560 by goroutine 32:
  github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).Close()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:712 +0x297
  github.com/cockroachdb/cockroach/pkg/storage.TestRaftSSTableSideloadingUpdatedReplicaID()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_sideload_test.go:1048 +0x7ef
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 43 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:186 +0xba
  github.com/cockroachdb/cockroach/pkg/storage/compactor.(*Compactor).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/compactor/compactor.go:110 +0x13c
  github.com/cockroachdb/cockroach/pkg/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1460 +0xa05
  github.com/cockroachdb/cockroach/pkg/storage.(*testContext).StartWithStoreConfig()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_test.go:208 +0x1333
  github.com/cockroachdb/cockroach/pkg/storage.(*testContext).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_test.go:158 +0x29d
  github.com/cockroachdb/cockroach/pkg/storage.TestRaftSSTableSideloadingUpdatedReplicaID()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_sideload_test.go:1004 +0x1f7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 32 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:978 +0x2cd
  github.com/cockroachdb/cockroach/pkg/storage_test.TestMain()
      /go/src/github.com/cockroachdb/cockroach/pkg/storage/main_test.go:57 +0x2a8
  main.main()
      _testmain.go:1062 +0x22a
==================

```

Please assign, take a look and update the issue accordingly.

--- a/pkg/storage/compactor/compactor.go
+++ b/pkg/storage/compactor/compactor.go
@@ -107,60 +107,64 @@ func (c *Compactor) Start(ctx context.Context, stopper *stop.Stopper) {
 	// started (this isn't great, but it's how it is right now).
 	c.poke()
 
-	stopper.RunWorker(ctx, func(ctx context.Context) {
-		var timer timeutil.Timer
-		defer timer.Stop()
+	// Run the Worker in a Task because the worker holds on to the engine and
+	// may still access it even though the stopper has allowed it to close.
+	_ = stopper.RunTask(ctx, "compactor", func(ctx context.Context) {
+		stopper.RunWorker(ctx, func(ctx context.Context) {
+			var timer timeutil.Timer
+			defer timer.Stop()
 
-		// The above timer will either be on c.minInterval() or c.maxAge(). The
-		// former applies if we know there are new suggestions waiting to be
-		// inspected: we want to look at them soon, but also want to make sure
-		// "related" suggestions arrive before we start compacting. When no new
-		// suggestions have been made since the last inspection, the expectation
-		// is that all we have to do is clean up any previously skipped ones (at
-		// least after sufficient time has passed), and so we wait out the max age.
-		var isFast bool
+			// The above timer will either be on c.minInterval() or c.maxAge(). The
+			// former applies if we know there are new suggestions waiting to be
+			// inspected: we want to look at them soon, but also want to make sure
+			// "related" suggestions arrive before we start compacting. When no new
+			// suggestions have been made since the last inspection, the expectation
+			// is that all we have to do is clean up any previously skipped ones (at
+			// least after sufficient time has passed), and so we wait out the max age.
+			var isFast bool
 
-		for {
-			select {
-			case <-stopper.ShouldStop():
-				return
+			for {
+				select {
+				case <-stopper.ShouldStop():
+					return
 
-			case <-c.ch:
-				// A new suggestion was made. Examine the compaction queue,
-				// which returns the number of bytes queued.
-				if bytesQueued, err := c.examineQueue(ctx); err != nil {
-					log.Warningf(ctx, "failed check whether compaction suggestions exist: %s", err)
-				} else if bytesQueued > 0 {
-					log.VEventf(ctx, 3, "compactor starting in %s as there are suggested compactions pending", c.minInterval())
-				} else {
-					// Queue is empty, don't set the timer. This can happen only at startup.
-					break
-				}
-				// Set the wait timer if not already set.
-				if !isFast {
+				case <-c.ch:
+					// A new suggestion was made. Examine the compaction queue,
+					// which returns the number of bytes queued.
+					if bytesQueued, err := c.examineQueue(ctx); err != nil {
+						log.Warningf(ctx, "failed check whether compaction suggestions exist: %s", err)
+					} else if bytesQueued > 0 {
+						log.VEventf(ctx, 3, "compactor starting in %s as there are suggested compactions pending", c.minInterval())
+					} else {
+						// Queue is empty, don't set the timer. This can happen only at startup.
+						break
+					}
+					// Set the wait timer if not already set.
+					if !isFast {
+						isFast = true
+						timer.Reset(c.minInterval())
+					}
+
+				case <-timer.C:
+					timer.Read = true
+					ok, err := c.processSuggestions(ctx)
+					if err != nil {
+						log.Warningf(ctx, "failed processing suggested compactions: %s", err)
+					}
+					if ok {
+						// The queue was processed, so either it's empty or contains suggestions
+						// that were skipped for now. Revisit when they are certainly expired.
+						isFast = false
+						timer.Reset(c.maxAge())
+						break
+					}
+					// More work to do, revisit after minInterval. Note that basically
+					// `ok == (err == nil)` but this refactor is left for a future commit.
 					isFast = true
 					timer.Reset(c.minInterval())
 				}
-
-			case <-timer.C:
-				timer.Read = true
-				ok, err := c.processSuggestions(ctx)
-				if err != nil {
-					log.Warningf(ctx, "failed processing suggested compactions: %s", err)
-				}
-				if ok {
-					// The queue was processed, so either it's empty or contains suggestions
-					// that were skipped for now. Revisit when they are certainly expired.
-					isFast = false
-					timer.Reset(c.maxAge())
-					break
-				}
-				// More work to do, revisit after minInterval. Note that basically
-				// `ok == (err == nil)` but this refactor is left for a future commit.
-				isFast = true
-				timer.Reset(c.minInterval())
 			}
-		}
+		})
 	})
 }
 

